Release process docs (#2214)
psf/black: docs/maintenance/release_process.md

# Release process

> _Black_ has had a lot of work automationg it's release process. This document sets out

Reviewer: ```suggestion
_Black_ has had a lot of work automating its release process. This document sets out
```

> to explain what everything does and how to relase _Black_ using said automation.

Reviewer: ```suggestion
to explain what everything does and how to release _Black_ using said automation.
```

## Cutting a Relase

To cut a realease, you must be a _Black_ maintainer with `GitHub Release` creation
access. Using this access, the release process is:

1. Cut a new PR editing `CHANGES.md` to version the latest changes
   1. Example PR: https://github.com/psf/black/pull/2192

>    2. Examele Title: `Update CHANGES.md for XX.X release`

Reviewer: ```suggestion
   2. Example title: `Update CHANGES.md for XX.X release`
```

   description of the GitHub Release.
4. Go and [cut a release](https://github.com/psf/black/releases) using the GitHub UI so
   that all workflows noted below are triggered.
   1. The release version and tag should be the [CalVer](https://calver.org) version
      _Black_ used for the current release e.g. `21.6` / `21.5b1`
   2. _Black_ uses [setuptools scm](https://pypi.org/project/setuptools-scm/) to pull
      the current version for the package builds and release.
5. Once the release is cut, you're basically done. It's a good practice to go and watch
   to make sure all the [GitHub Actions](https://github.com/psf/black/actions) pass,
   although you should receive an email to your registered GitHub email address should
   one fail.
   1. You should see all the release workflows and lint/unittests workflows running on
      the new tag in the Actions UI

If anything fails, please go read the respective action's log output and configuration
file to reverse engineer your way to a fix/soluton.

## Release workflows

All _Blacks_'s automation workflows use GitHub Actions. All workflows are therefore

> configured using `.yml` files in `.github/workflows` directory of the _Black_
> repositroy.

Reviewer: ```suggestion
configured using `.yml` files in the `.github/workflows` directory of the _Black_
repository.
```

Below are descriptions of our release workflows.

### Docker

This workflow uses the QEMU powered `buildx` feature of docker to upload a `arm64` and
`amd64`/`x86_64` build of the official _Black_ docker image™.

- Currently this workflow uses an API Token associated with @cooperlees account

### pypi_upload

This workflow builds a Python
[sdist](https://docs.python.org/3/distutils/sourcedist.html) and
[wheel](https://pythonwheels.com) using the latest
[setuptools](https://pypi.org/project/setuptools/)

> w/[wheel](https://pypi.org/project/wheel/) modules.

Reviewer: ```suggestion
and [wheel](https://pypi.org/project/wheel/) modules.
```

It will then use [twine](https://pypi.org/project/twine/) to upload both release formats

> to PyPI for general downloading of the _Black_ python package. This is where

Reviewer: ```suggestion
to PyPI for general downloading of the _Black_ Python package. This is where
```

[pip](https://pypi.org/project/pip/) looks by default.

> - Currently this workflow uses an API TOken associated with @ambv PyPI account

Reviewer: ```suggestion
- Currently this workflow uses an API token associated with @ambv's PyPI account
```

### Upload self-contained binaries

> This workflow builds self container binaries for multiple platforms. This allows people

Reviewer: ```suggestion
This workflow builds self-contained binaries for multiple platforms. This allows people
```

Author: LoL @ this one. Such a me typo.
